### PR TITLE
nixos/go-neb: init

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

Go-neb is a matrix bot that was packaged in #87854. Now we deliver a simple module to easily set this up.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

@tilpner @maralorn

### CODE REVIEW

File: nixos/modules/services/networking/go-neb.nix
Comment (aanderse) [inline nixos/modules/services/networking/go-neb.nix:41]: Bonus points for using the [systemd.services.\<name\>.environment](https://nixos.org/nixos/options.html#systemd.+services.%3Cname%3E.environment) option.

File: nixos/modules/services/networking/go-neb.nix
Comment (aanderse) [inline nixos/modules/services/networking/go-neb.nix:24]: Any interest in providing a mechanism for handling secrets in here? From the looks of the format likely not trivial to add great support...

**Review by aanderse — COMMENTED** [review]
Nice and simple module :+1:

Even more bonus points if it comes with a NixOS test :sparkles:

**Comment by maralorn** [comment]
I have copied this module to my server and have alertmanager warnings (and a wikipedia bot) up and running with this.
```nix
{
  services.go-neb = {
    enable = true;
    baseUrl = "http://localhost";
    config = {
      clients = [{
        UserId = "@marabot:maralorn.de";
        AccessToken = config.m-0.private.matrix_marabot_token;
        HomeServerUrl = "https://matrix.maralorn.de";
        Sync = true;
        AutoJoinRooms = true;
        DisplayName = "marabot";
      }];
      realms = [ ];
      sessions = [ ];
      services = [
        {
          ID = "alertmanager_service";
          Type = "alertmanager";
          UserId = "@marabot:maralorn.de";
          Config = {
            webhook_url =
              "http://localhost:4050/services/hooks/YWxlcnRtYW5hZ2VyX3NlcnZpY2UK";
            rooms = {
              "!negVsngnYOmXYCLKiO:maralorn.de" = {
                text_template = ''
                  {{range .Alerts -}} [{{ .Status }}] {{index .Labels "alertname" }}: {{index .Annotations "description"}} {{ end -}}'';
                  html_template = ''
                    {{range .Alerts -}}
                      {{ $severity := index .Labels "severity" }}
                      {{ if eq .Status "firing" }}
                        {{ if eq $severity "critical"}}
                          <font color='red'><b>[FIRING - CRITICAL]</b></font>
                        {{ else if eq $severity "warning"}}
                          <font color='orange'><b>[FIRING - WARNING]</b></font>
                        {{ else }}
                          <b>[FIRING - {{ $severity }}]</b>
                        {{ end }}
                      {{ else }}
                        <font color='green'><b>[RESOLVED]</b></font>
                      {{ end }}
                      {{ index .Labels "alertname"}}:{{ index .Annotations "description"}}
                      <a href="{{ .GeneratorURL }}">source</a><br/>
                    {{end -}}
                    '';
                msg_type = "m.text"; # Must be either `m.text` or `m.notice`
              };
            };
          };
        }
        {
          ID = "wikipedia_service";
          Type = "wikipedia";
          UserID = "@marabot:maralorn.de"; # requires a Syncing client
          Config = { };
        }
      ];
    };
  };
}
```

File: nixos/modules/services/networking/go-neb.nix
Comment (mweinelt) [reply nixos/modules/services/networking/go-neb.nix]: Sure, thanks for the pointer.

File: nixos/modules/services/networking/go-neb.nix
Comment (mweinelt) [reply nixos/modules/services/networking/go-neb.nix:24]: The only secret worth protecting would be `clients.*.AccessToken`. I'm not sure how I would keep that out of the nix store when the program does not support reading secrets from dedicated files.

Comment (aanderse) [reply nixos/modules/services/networking/go-neb.nix:24]: Yeah the solution wouldn't be pretty. Some modules will read the password on startup and then copy+paste+modify the generated configuration file replacing any secrets in a protected directory. Using `DynamicUser` adds additional complications. Secrets management in NixOS is not an easy thing, unfortunately.

**Comment by mweinelt** [comment]
Simple test added.

File: nixos/modules/services/networking/go-neb.nix
Comment (aanderse) [inline nixos/modules/services/networking/go-neb.nix:40]: Quotes no longer required here.

Comment (mweinelt) [reply nixos/modules/services/networking/go-neb.nix]: Indeed!

**Comment by tilpner** [comment]
In https://github.com/matrix-org/go-neb#quick-start, DATABASE_{TYPE,URL} are set, which is not replicated here. Where does this module store its state?

File: nixos/modules/services/networking/go-neb.nix
Comment (tilpner) [inline nixos/modules/services/networking/go-neb.nix:25]: Multiple definitions of attrs are apparently not merged properly. [According to @Infinisil, `types.uniq types.attr` might be appropriate](https://logs.nix.samueldr.com/nixos-dev/2020-06-02#3549491;) to prevent that merging.

**Comment by maralorn** [comment]
@tilpner the bot has two modes. Statefully saving the config in an sqlite DB or reading the config from a config file. So the DB you are mentioning is not data state but only for config state. We opted to support only the second option, because it supports the same features and is much more nix like.

File: nixos/modules/services/networking/go-neb.nix
Comment (maralorn) [reply nixos/modules/services/networking/go-neb.nix:25]: That mistake is probably all over nixpkgs right? But seems easy enough to fix here.

**Comment by mweinelt** [comment]
> @tilpner the bot has two modes. Statefully saving the config in an sqlite DB or reading the config from a config file. So the DB you are mentioning is not data state but only for config state. We opted to support only the second option, because it supports the same features and is much more nix like.

The partial source for this statement is here: https://github.com/matrix-org/go-neb/blob/81842c69f831b7dccaa685ccacbbe218d059d76c/config.sample.yaml#L5-L9

File: nixos/modules/services/networking/go-neb.nix
Comment (mweinelt) [reply nixos/modules/services/networking/go-neb.nix:25]: In fact i don't see that used anywhere. Wild!

**Comment by tilpner** [comment]
I see. It surprised me that they were exclusive modes, I was hoping it would operate in a hybrid mode, as e.g. [reliable RSS feed consumption requires state](https://github.com/matrix-org/go-neb/blob/81842c69f831b7dccaa685ccacbbe218d059d76c/services/rssbot/rssbot.go#L97-L98).

Supporting only the declarative mode in this module is reasonable though. Thanks for writing it! :)

**Review by tilpner — APPROVED** [review]

**Comment by maralorn** [comment]
@aanderse Are you also satisfied with the state of this?

**Review by aanderse — APPROVED** [review]
Yeah this is a winner in my books. Thanks for the extra work on this one. Bonus points for all! :sparkles:

I'm not a subject matter expert so I wouldn't think to merge this... but if you're just waiting for a merge and everyone agrees I'm certainly happy to do so based on that :+1:

**Comment by aanderse** [comment]
... and just as I say that I notice the test isn't added to the list of tests. Please add and we're good.

**Comment by mweinelt** [comment]
Good catch. Added here: https://github.com/NixOS/nixpkgs/pull/89327/files#diff-8e2944fa8b75f34d13428287028e1323R122

Thanks!

**Review by maralorn — APPROVED** [review]

**Comment by maralorn** [comment]
@aanderse Can you maybe point us to a subject matter expert? Otherwise I'd opt for the merge.

**Comment by aanderse** [comment]
@maralorn maybe @Ma27 knows a thing or two... but this looks really good, so I'll merge by end of day if no one has any objections.

Thanks! :tada:

**Comment by aanderse** [comment]
@GrahamcOfBorg test go-neb

**Comment by aanderse** [comment]
Oops! Thanks team :heart:

**Pull request merged**
